src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,792 @@
+            {
+                _semanticModel = semanticModel;
+                _semanticFacts = semanticFacts;
+                _cancellationToken = cancellationToken;
+            }
+
+            public bool TryConvert(QueryExpressionSyntax source, out DocumentUpdateInfo documentUpdateInfo)
+            {
+                if (source.DescendantTrivia().Any(trivia => trivia.MatchesKind(
+                        SyntaxKind.SingleLineCommentTrivia,
+                        SyntaxKind.MultiLineCommentTrivia,
+                        SyntaxKind.MultiLineDocumentationCommentTrivia) ||
+                    source.ContainsDirectives))
+                {
+                    documentUpdateInfo = default;
+                    return false;
+                }
+
+                // GetDiagnostics is expensive. Move it to the end if there were no bail outs from the algorithm.
+                // TODO likely adding more semantic checks will perform checks we expect from GetDiagnostics
+                // We may consider removing GetDiagnostics.
+                // https://github.com/dotnet/roslyn/issues/25639
+                if (!TryConvertInternal(source, out documentUpdateInfo) ||
+                    _semanticModel.GetDiagnostics(source.Span, _cancellationToken).Any())
+                {
+                    documentUpdateInfo = default;
+                    return false;
+                }
+
+                return true;
+            }
+
+            private bool TryConvertInternal(QueryExpressionSyntax source, out DocumentUpdateInfo documentUpdateInfo)
+            {
+                // (from a in b select a); 
+                SyntaxNode parent = source.WalkUpParentheses().Parent;
+
+                switch (parent.Kind())
+                {
+                    // return from a in b select a;
+                    case SyntaxKind.ReturnStatement:
+                        return TryConvertIfInReturnStatement(source, (ReturnStatementSyntax)parent, out documentUpdateInfo);
+                    // foreach(var x in from a in b select a)
+                    case SyntaxKind.ForEachStatement:
+                        return TryConvertIfInForEach(source, (ForEachStatementSyntax)parent, out documentUpdateInfo);
+                    // (from a in b select a).ToList(), (from a in b select a).Count(), etc.
+                    case SyntaxKind.SimpleMemberAccessExpression:
+                        return TryConvertIfInMemberAccessExpression(source, (MemberAccessExpressionSyntax)parent, out documentUpdateInfo);
+                    // var a = new [] { from a in b select a };                    
+                    case SyntaxKind.ArrayInitializerExpression:
+                    // var a = from a in b select a
+                    case SyntaxKind.EqualsValueClause:
+                    // a = from a in b select a
+                    case SyntaxKind.SimpleAssignmentExpression:
```

Comment:
>SyntaxKind.SimpleAssignmentExpression [](start = 25, length = 37)

I would check that the query is at the right position, i.e. on the right. I can probably come up with a valid code where the query is on the left.

Reply:
I have added a check. Could you please remind me an example with the valid code is on the left?

---
In reply to: [176928453](https://github.com/dotnet/roslyn/pull/25362#discussion_r176928453) [](ancestors = 176928453)

Reply:
> I have added a check. Could you please remind me an example with the valid code is on the left?

For example, the last query operator  returns ```ref``` something. One can assign through that ref 

---
In reply to: [177826396](https://github.com/dotnet/roslyn/pull/25362#discussion_r177826396) [](ancestors = 177826396,176928453)